themes/default/content/blog/testing-in-practice/index.md

```diff
@@ -0,0 +1,171 @@
+---
+title: "Infrastructure Testing in Practice"
+date: 2021-07-23
+meta_desc: "Infrastructure testing is core to cloud engineering; this article examines unit, property, and integration testing using infrastructure as code."
+meta_image: testing-in-practice.png
+authors:
+    - sophia-parafina
+tags:
+    - testing
+    - cloud engineering
+---
+
+In the [previous article]({{< relref "/blog/infrastructure-testing-concepts">}}) we discussed how to apply software testing methodologies to cloud engineering. We also examined testing regimes starting from the testing pyramid to the trophy and honeycomb models of testing better suited to distributed and cloud architectures. These testing regimes include three types of tests suited for cloud architectures:
+
+- unit tests for testing methods and functions within a service
+- property tests for validating specified service outputs
+- integration tests to ensure that resources interact as specified
+
+In this article, we'll do a deep dive into each of these testing methods.
+
+<!--more-->
+
+Cloud engineering testing differs from application testing becausemultiple dependencies between components can have high latency. Let's examine these dependencies using an [example application](https://github.com/pulumi/examples/tree/master/aws-stackreference-architecture) composed of an application layer, a database layer, and a network layer. We can visualize these dependencies with Pulumi's resource graph.The graph is constructed from the [inputs]({{< relref "/docs/intro/concepts/inputs-outputs" >}}) needed to create a resource.
+
+The resource graph for the application layer shows many dependencies. The application deploys containers in AWS Fargate, which are instantiated with an application service that requires a service definition with many inputs.
+
+![Resource graph](./image2.png)
+
+Unit tests are ideal for validating inputs and ensuring that cloud resources are created. Let's look at a simple example with just two resources, an EC2 instance running a web server and a SecurityGroup policy.
+
+![Simple example](./image1.png)
+
+We can use mocks to test if the webserver has a name tag. First, we'll create the mocks and import the infrastructure that uses the mocks.
```

Comment:
do we want to link out to a definition of a mock? https://en.wikipedia.org/wiki/Mock_object

(there might be a better one to use)

we could assume the reader already has a familiarity with mocks of course. Just a suggestion and I'm not sure either way :)